File: auto2fa-mac/Auto2FA/Models/Tunnel.swift
import Foundation

/// Mirror of the daemon's tunnel snapshot.
struct Tunnel: Identifiable, Codable, Equatable, Hashable {
    let name: String
    let localPort: Int
    let remotePort: Int
    let jumpCandidates: [String]?
    let lastNode: String?
    let lastUser: String?
    let autoStart: Bool
    let postConnectCmd: String?
    let tags: [String]
    let urlPath: String?
    let activeJump: String?
    let status: String
    let lastMsg: String
    let lastAliveAt: Double
    let totalUptimeSec: Double
    let connectCount: Int
    let failCount: Int

    var id: String { name }
    var url: String { "localhost:\(localPort)" }

    /// Full browser URL: prepends http:// + appends user-defined urlPath
    /// (e.g. "/?token=abc") if set. Used by "Open in browser" and ⌘O.
    var browserURL: String {
        let p = urlPath?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let suffix = p.isEmpty ? "" : (p.hasPrefix("/") || p.hasPrefix("?") ? p : "/" + p)
        return "http://localhost:\(localPort)\(suffix)"
    }

    enum CodingKeys: String, CodingKey {
        case name, status, tags
        case localPort = "local_port"
        case remotePort = "remote_port"
        case jumpCandidates = "jump_candidates"
        case lastNode = "last_node"
        case lastUser = "last_user"
        case autoStart = "auto_start"
        case postConnectCmd = "post_connect_cmd"
        case urlPath = "url_path"
        case activeJump = "active_jump"
        case lastMsg = "last_msg"
        case lastAliveAt = "last_alive_at"
        case totalUptimeSec = "total_uptime_sec"
        case connectCount = "connect_count"
        case failCount = "fail_count"
    }

    // Defaults so older daemon snapshots still decode.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.name = try c.decode(String.self, forKey: .name)
        self.status = try c.decode(String.self, forKey: .status)
        self.localPort = try c.decode(Int.self, forKey: .localPort)
        self.remotePort = try c.decode(Int.self, forKey: .remotePort)
        self.jumpCandidates = try c.decodeIfPresent([String].self, forKey: .jumpCandidates)
        self.lastNode = try c.decodeIfPresent(String.self, forKey: .lastNode)
        self.lastUser = try c.decodeIfPresent(String.self, forKey: .lastUser)
        self.autoStart = try c.decode(Bool.self, forKey: .autoStart)
        self.postConnectCmd = try c.decodeIfPresent(String.self, forKey: .postConnectCmd)
        self.tags = (try? c.decode([String].self, forKey: .tags)) ?? []
        self.urlPath = try? c.decodeIfPresent(String.self, forKey: .urlPath)
        self.activeJump = try c.decodeIfPresent(String.self, forKey: .activeJump)
        self.lastMsg = try c.decode(String.self, forKey: .lastMsg)
        self.lastAliveAt = (try? c.decode(Double.self, forKey: .lastAliveAt)) ?? 0
        self.totalUptimeSec = (try? c.decode(Double.self, forKey: .totalUptimeSec)) ?? 0
        self.connectCount = (try? c.decode(Int.self, forKey: .connectCount)) ?? 0
        self.failCount = (try? c.decode(Int.self, forKey: .failCount)) ?? 0
    }

    /// Human-friendly cumulative uptime (since daemon start).
    var uptimeHuman: String {
        let s = max(0, totalUptimeSec)
        if s < 60 { return "\(Int(s))s" }
        if s < 3600 { return "\(Int(s/60))m" }
        if s < 86400 { return String(format: "%.1fh", s/3600) }
        return String(format: "%.1fd", s/86400)
    }

    /// Human-friendly "alive 2h", "last alive 5m ago", "never alive".
    /// Negative numbers handled defensively even though they shouldn't happen.
    func aliveSince(now: Date = Date()) -> String? {
        guard lastAliveAt > 0 else { return nil }
        let elapsed = now.timeIntervalSince1970 - lastAliveAt
        let abs = max(0, elapsed)
        let prefix = (displayState == .alive) ? "alive" : "last alive"
        let suffix = (displayState == .alive) ? "" : " ago"
        let unit: String
        if abs < 60 { unit = "\(Int(abs))s" }
        else if abs < 3600 { unit = "\(Int(abs/60))m" }
        else if abs < 86400 { unit = "\(Int(abs/3600))h" }
        else { unit = "\(Int(abs/86400))d" }
        return "\(prefix) \(unit)\(suffix)"
    }

    enum DisplayState: String {
        case alive, starting, stale, idle, portBusy, failed, unknown

        init(from raw: String) {
            switch raw {
            case "alive": self = .alive
            case "starting": self = .starting
            case "stale": self = .stale
            case "idle": self = .idle
            case "port_busy": self = .portBusy
            case "failed": self = .failed
            default: self = .unknown
            }
        }
    }

    var displayState: DisplayState { DisplayState(from: status) }
}

File: auto2fa-rs/crates/a2fa-core/src/model/tunnel.rs
use serde::{Deserialize, Serialize};

use crate::model::status::TunnelStatus;

/// A snapshot of a single tunnel as emitted by `_tunnel_snapshot` in daemon.py,
/// augmented with the persisted fields from `TunnelState` in tunnels.py.
///
/// Ports are `u16` for direct serde compatibility with the JSON wire format.
/// Validation (1024..=65535) is enforced at creation boundaries (e.g. tunnel add),
/// not in this struct — keeping deserialization infallible for IPC payloads.
///
/// Timestamps (`last_alive_at`, `total_uptime_sec`) are `f64` (Unix epoch seconds)
/// matching Python's `time.time()` and `float` accumulator.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tunnel {
    // ---- Persisted fields (tunnels.json) ------------------------------------

    /// Unique tunnel name / identifier.
    pub name: String,

    /// Local port the SSH -L forward binds on 127.0.0.1.
    pub local_port: u16,

    /// Remote port on the compute node that local_port forwards to.
    pub remote_port: u16,

    /// Explicit set of jump-host names to try, in priority order.
    /// `None` means "use any ready host in passwords.json".
    pub jump_candidates: Option<Vec<String>>,

    /// Last compute node this tunnel targeted (e.g. "gpunode01").
    pub last_node: Option<String>,

    /// Last UNIX user used for the `user@node` part of the ssh command.
    pub last_user: Option<String>,

    /// When `Some(host)`, this tunnel forwards local_port → localhost:remote_port
    /// directly ON that registered host (`ssh -N -L … <host>`) — NO jump host and
    /// NO SLURM compute node. `None` = the default SLURM compute-node forward.
    ///
    /// `#[serde(default)]`: old tunnels.json and old IPC snapshots omit this field
    /// and must still decode (→ None = unchanged compute behavior).
    #[serde(default)]
    pub direct_host: Option<String>,

    /// If true, the tunnel is automatically started at daemon boot.
    pub auto_start: bool,

    /// Optional `/bin/sh -c` command run when the tunnel first reaches "alive".
    pub post_connect_cmd: Option<String>,

    /// User-defined grouping tags (stored verbatim).
    pub tags: Vec<String>,

    /// Optional URL path/query suffix for "Open in browser" (e.g. `"?token=abc"`).
    pub url_path: Option<String>,

    /// Whether the user wants this tunnel alive right now.
    /// Persisted so auto-recovery survives daemon restarts.
    ///
    /// `#[serde(default)]`: the IPC `tunnel_snapshot` deliberately OMITS this
    /// field (clients shouldn't depend on it), but the TUI deserializes
    /// snapshots into this very struct — without the default, every
    /// `list_tunnels` decode failed on "missing field wants_alive" and the
    /// TUI's tunnel pane stayed permanently empty.
    #[serde(default)]
    pub wants_alive: bool,

    // ---- Runtime / snapshot fields ------------------------------------------

    /// Current lifecycle status of the tunnel.
    pub status: TunnelStatus,

    /// Name of the jump host currently forwarding this tunnel, if alive.
    pub active_jump: Option<String>,

    /// Human-readable last status message.
    pub last_msg: String,

    /// Unix timestamp (seconds) of the last time this tunnel reached "alive".
    /// `0.0` means it has never been alive.
    pub last_alive_at: f64,

    /// Cumulative seconds spent in "alive" state (live-computed in daemon.py
    /// via `_live_uptime` to include the current run).
    pub total_uptime_sec: f64,

    /// Number of successful "alive" transitions since daemon start.
    pub connect_count: u32,

    /// Number of failed/stale transitions since daemon start.
    pub fail_count: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A tunnel JSON WITHOUT direct_host must decode (→ None) so old
    /// tunnels.json / old snapshots keep working.
    #[test]
    fn direct_host_defaults_to_none_when_absent() {
        let json = r#"{
            "name": "nb", "local_port": 8888, "remote_port": 8888,
            "jump_candidates": null, "last_node": null, "last_user": null,
            "auto_start": false, "post_connect_cmd": null, "tags": [],
            "url_path": null, "status": "idle", "active_jump": null,
            "last_msg": "", "last_alive_at": 0.0, "total_uptime_sec": 0.0,
            "connect_count": 0, "fail_count": 0
        }"#;
        let t: Tunnel = serde_json::from_str(json).expect("decode without direct_host");
        assert_eq!(t.direct_host, None);
    }

    /// A direct tunnel round-trips its host.
    #[test]
    fn direct_host_round_trips() {
        let json = r#"{
            "name": "web", "local_port": 9000, "remote_port": 9000,
            "jump_candidates": null, "last_node": null, "last_user": null,
            "auto_start": false, "post_connect_cmd": null, "tags": [],
            "url_path": null, "direct_host": "loginhost", "status": "idle",
            "active_jump": null, "last_msg": "", "last_alive_at": 0.0,
            "total_uptime_sec": 0.0, "connect_count": 0, "fail_count": 0
        }"#;
        let t: Tunnel = serde_json::from_str(json).expect("decode with direct_host");
        assert_eq!(t.direct_host.as_deref(), Some("loginhost"));
        let back = serde_json::to_value(&t).unwrap();
        assert_eq!(back["direct_host"], "loginhost");
    }
}
